src/compiler/checker.ts

```diff
@@ -41544,8 +41540,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             //   and Expr must be an expression of type Any, an object type, or a type parameter type.
             const varExpr = node.initializer;
             const leftType = checkExpression(varExpr);
-            if (varExpr.kind === SyntaxKind.ArrayLiteralExpression || varExpr.kind === SyntaxKind.ObjectLiteralExpression) {
-                error(varExpr, Diagnostics.The_left_hand_side_of_a_for_in_statement_cannot_be_a_destructuring_pattern);
+            if (isAssignmentPattern(varExpr)) {
+                checkDestructuringAssignment(varExpr, getIndexTypeOrString(rightType));

```

Comment:
I'm not sure if this should actually be using `getIndexTypeOrString(rightType)`. I used it because the branch below this uses that but because it does use it this code typechecks but, IMHO, it shouldn't ([TS playground](https://www.typescriptlang.org/play?target=99&ts=5.1.3#code/GYVwdgxgLglg9mABFApgZygCjgIwFYBciA3ogIZECMA3IjkQEyIC+AlCQFCKIA2KUiANYoAnkQBEZcYgA+icTnHUuiYHABOiTMJGIYSXHnbEVzDmaA)): 
```ts
function test(obj: { a: 1; b: 2 }) {
  let key: "a" | "b";
  for (key in obj) {
  }
}
```

Reply:
Created this PR to address/discuss this: https://github.com/microsoft/TypeScript/pull/54856